tools/gengen/util/generator.go

```diff
@@ -0,0 +1,651 @@
+	"github.com/filecoin-project/go-filecoin/internal/pkg/cborutil"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/consensus"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/crypto"
+	e "github.com/filecoin-project/go-filecoin/internal/pkg/enccid"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/encoding"
+	gfcstate "github.com/filecoin-project/go-filecoin/internal/pkg/state"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/types"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/actor"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/state"
+)
+
+type cstore struct {
+	ctx context.Context
+	cbor.IpldStore
+}
+
+func (s *cstore) Context() context.Context {
+	return s.ctx
+}
+
+var (
+	defaultAccounts map[address.Address]abi.TokenAmount
+)
+
+func init() {
+	defaultAccounts = map[address.Address]abi.TokenAmount{
+		builtin.RewardActorAddr:     abi.NewTokenAmount(10000000000),
+		builtin.BurntFundsActorAddr: abi.NewTokenAmount(0),
+	}
+}
+
+type GenesisGenerator struct {
+	// actor state
+	stateTree state.Tree
+	store     vm.Storage
+	cst       cbor.IpldStore
+	vm        consensus.GenesisVM
+
+	keys      []*crypto.KeyInfo
+	pnrg      *mrand.Rand
+	chainRand crypto.ChainRandomnessSource
+	cfg       *GenesisCfg
+}
+
+func NewGenesisGenerator(bs blockstore.Blockstore) *GenesisGenerator {
+	cst := cborutil.NewIpldStore(bs)
+	g := GenesisGenerator{}
+	g.stateTree = state.NewState(cst)
+	g.store = vm.NewStorage(bs)
+	g.vm = vm.NewVM(g.stateTree, &g.store).(consensus.GenesisVM)
+	g.cst = cst
+
+	g.chainRand = crypto.ChainRandomnessSource{Sampler: &crypto.GenesisSampler{TicketBytes: consensus.GenesisTicket.VRFProof}}
+	return &g
+}
+
+func (g *GenesisGenerator) Init(cfg *GenesisCfg) error {
+	g.pnrg = mrand.New(mrand.NewSource(cfg.Seed))
+	keys, err := genKeys(cfg.Keys, g.pnrg)
+	if err != nil {
+		return err
+	}
+	g.keys = keys
+	g.cfg = cfg
+
+	return nil
+}
+
+func (g *GenesisGenerator) flush(ctx context.Context) (cid.Cid, error) {
+	err := g.store.Flush()
+	if err != nil {
+		return cid.Undef, err
+	}
+	return g.stateTree.Commit(ctx)
+}
+
+func (g *GenesisGenerator) createActor(addr address.Address, codeCid cid.Cid, balance abi.TokenAmount, stateFn func() (interface{}, error)) (*actor.Actor, error) {
+	a := actor.Actor{
+		Code:       e.NewCid(codeCid),
+		CallSeqNum: 0,
+		Balance:    balance,
+	}
+	if stateFn != nil {
+		state, err := stateFn()
+		if err != nil {
+			return nil, fmt.Errorf("failed to create state")
+		}
+		headCid, err := g.store.Put(state)
+		if err != nil {
+			return nil, fmt.Errorf("failed to store state")
+		}
+		a.Head = e.NewCid(headCid)
+	}
+	if err := g.stateTree.SetActor(context.Background(), addr, &a); err != nil {
+		return nil, fmt.Errorf("failed to create actor during genesis block creation")
+	}
+
+	return &a, nil
+}
+
+func (g *GenesisGenerator) setupDefaultActors(ctx context.Context) error {
```

Comment:
This is copypasted from consensus/genesis.go.  It 100% belongs here and consensus/genesis.go utils should either be deleted entirely or become a thin wrapper around gengen.  Right now I'm deferring that move to another PR but will do it here if reviewers prefer.